### ISSUE TITLE
feature request: keep samtools fasta memory usage below 100MB

### ISSUE THREAD

**Issue by tommycarstensen** [samtools/htslib#502]
I ran this process:

    $samtools fasta $cram | bgzip > $fasta

It used slightly more than 100MB of memory. Would it be possible to memory optimise it? It probably shouldn't be using more than 10-50MB of RAM. Thanks.

**Comment by jkbonfield** [comment]
For CRAM it's probably dependent on how long the reference it loaded was.

It may be possible to reduce this via mmap (it's an option, but not sure if it is compiled in by default).  I'll investigate.

**Comment by mp15** [comment]
I was wondering whether we should strongly suggest in our documentation that when sorting by name reference based compression be turned off? It seems likely to send us seeking all over the reference as adjacent reads are unlikely to be mapped to proximal locations within the genome.

**Comment by tommycarstensen** [comment]
Thanks @jkbonfield ! Please make it a low priority. And please ignore the feature request entirely, if it's in any way difficult to implement. The memory usage just seemed a bit excessive to me. Thanks!

**Comment by jkbonfield** [comment]
Good point @mp15.  Is this a name sorted file? 

Also I can confirm there is a difference whether using the MD5 reference cache or disabling this (clear REF_PATH environ) and supplying a --reference ref.fa arg.  The md5 method is far far lower in memory, but I can't quite pin down why yet.   The entire reference handling in htslib/CRAM is a bit of a nightmare frankly!

**Comment by mp15** [comment]
Any file going into ```samtools fasta``` should be name sorted.

**Comment by jkbonfield** [comment]
Even if it's not paired end?  Anyway I agree that any name sorted data should not be using the reference as it's horribly slow once you do that and inherently has to start loading the entire reference into memory to avoid it becoming I/O bound.

**Comment by mp15** [comment]
Yes even if it's not paired end they should have at least shuffled it, because otherwise your insert size estimation going into your aligner (vast majority of the use case) will be horribly biased.

**Comment by jkbonfield** [comment]
@tommycarstensen Is your file using reference based compression still?

You can try turning this off via:

```
samtools view -O cram,no_ref $in.cram -o $out.cram
```

Do you find memory usage reduced for ```$out.cram``` over ```$in.cram```?

I'm wondering whether we should make the default usage of reference auto-tune to the input @HD sort order.  Perhaps with a new ```-O cram,use_ref``` option to force it to be on when people really need high compression of name sorted data (at the expense of memory / cpu).

**Comment by tommycarstensen** [comment]
@jkbonfield I'm not too sure which kind of sorting is applied to the cram.

Input cram:

    samtools view /lustre/scratch115/projects/adrp1/users/tc9/RNAseq/cram/21758_1#1.cram | cut -f1 | head
    HS31_21758:1:1104:14863:34672
    HS31_21758:1:1104:14863:34672
    HS31_21758:1:1104:14863:34672
    HS31_21758:1:1104:14863:34672
    HS31_21758:1:1104:14863:34672
    HS31_21758:1:2106:14863:4214
    HS31_21758:1:1107:3287:88540
    HS31_21758:1:1107:3287:88540
    HS31_21758:1:1107:3287:88540
    HS31_21758:1:1107:3287:88540

Output fasta:

    zcat /lustre/scratch115/projects/adrp1/users/tc9/RNAseq/fasta/21758_1#1.fasta.gz | head -n20 | grep ^">"
    >HS31_21758:1:2302:1686:88341/1
    >HS31_21758:1:2108:1340:63461/1
    >HS31_21758:1:1302:14159:87170/1
    >HS31_21758:1:1302:14159:87170/2
    >HS31_21758:1:2102:21023:51608/1
    >HS31_21758:1:2311:20417:80392/1
    >HS31_21758:1:2202:17010:50940/1
    >HS31_21758:1:2311:20417:80392/2
    >HS31_21758:1:2203:8145:30792/1
    >HS31_21758:1:2203:8145:30792/2

I did this:

    samtools view -O cram,no_ref $in.cram -o $out.cram

I tried using `$in.cram` and `$out.cram` as input for samtools fasta. This was one of the cram files that required less than 100MB. `$in.cram` used `61MB` and `$out.cram` used `20MB` (including a pipe to `bgzip`).

**Comment by jkbonfield** [comment]
They don't look sorted to me.  Specifically the output fasta is /1, /1, /1, /2, /1, /1, etc.  This is likely not in a format usable by most tools.

@mp15,  is this your function?  If so do you have time to update it to check header sort order (sorted by name or unordered and grouped by name), plus update the documentation too?  The man page makes no reference to the sort order.

As far as memory goes, I think this shows the benefit of not using a reference.  (You'd find with name sorted data that actively using a reference makes the memory explode to 3Gb or so due to keeping the entire reference in memory to handle the now-random access pattern.)  I should update htslib so that name sorted data by default doesn't do reference based compression.
